File: buildSrc/src/test/java/com/fizzpod/gradle/plugins/layout/LayoutPathSpec.java
/* (C) 2024 */
/* SPDX-License-Identifier: Apache-2.0 */
package com.fizzpod.gradle.plugins.layout;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

import static org.junit.jupiter.api.Assertions.*;

public class LayoutPathSpec {

    @TempDir
    Path temporaryFolderPath;

    private File getTempFolder() {
        return temporaryFolderPath.toFile();
    }

    @Test
    void createPathInRootFolder() {
        // when:
        LayoutPath layoutPath = new LayoutPath("folder1");

        // then:
        layoutPath.write(getTempFolder());

        // expect:
        File folder = new File(getTempFolder(), "folder1");
        assertTrue(folder.exists(), "Folder should exist");
        assertTrue(folder.isDirectory(), "Path should be a directory");
    }

    @Test
    void createPathWithForwardSlashAtStart() {
        // when:
        LayoutPath layoutPath = new LayoutPath("/folder1");

        // then:
        layoutPath.write(getTempFolder());

        // expect:
        // LayoutPath.java handles relativizing paths starting with "/"
        File folder = new File(getTempFolder(), "folder1");
        assertTrue(folder.exists(), "Folder should exist");
        assertTrue(folder.isDirectory(), "Path should be a directory");
    }

    @Test
    void createPathWithMultipleSubfolders() {
        // when:
        LayoutPath layoutPath = new LayoutPath("/folder1/folder1_1/folder1_1_1");

        // then:
        layoutPath.write(getTempFolder());

        // expect:
        File folder = new File(getTempFolder(), "folder1/folder1_1/folder1_1_1");
        assertTrue(folder.exists(), "Folder structure should exist");
        assertTrue(folder.isDirectory(), "Path should be a directory");
    }

    @Test
    void createPathWithFileInTheWay() throws IOException {
        // setup:
        File existingFile = new File(getTempFolder(), "file1");
        assertTrue(existingFile.createNewFile(), "Setup: Failed to create file1");
        Files.writeString(existingFile.toPath(), "content");


        // when:
        LayoutPath layoutPath = new LayoutPath("/file1"); // Attempting to create a directory named "file1"

        // then:
        layoutPath.write(getTempFolder()); // LayoutPath.java's write should not create a dir if a file exists

        // expect:
        assertTrue(existingFile.exists(), "Existing file should still exist");
        assertTrue(existingFile.isFile(), "Path should remain a file");
        // Ensure it didn't somehow become a directory
        long size = Files.size(existingFile.toPath()); // Reading size to ensure it's still the file
        assertEquals("content".length(), (int)size, "File content should be intact");
    }

    @Test
    void createPathWithFileInSubFolderPosition() throws IOException {
        // setup:
        File blockingFile = new File(getTempFolder(), "file1");
        assertTrue(blockingFile.createNewFile(), "Setup: Failed to create blocking file 'file1'");

        // when:
        // Trying to create a directory structure where "file1" is a file, not a directory
        LayoutPath layoutPath = new LayoutPath("/file1/folder1_1");

        // then:
        layoutPath.write(getTempFolder()); // This should fail to create "folder1_1" inside "file1"

        // expect:
        assertTrue(blockingFile.exists(), "Existing file 'file1' should still exist");
        assertTrue(blockingFile.isFile(), "Existing 'file1' should remain a file");

        // The Spock test had "file1/folder1", but LayoutPath input was "/file1/folder1_1"
        File newFolder = new File(getTempFolder(), "file1/folder1_1");
        assertFalse(newFolder.exists(), "New folder 'file1/folder1_1' should not be created");
    }
}

File: buildSrc/src/test/kotlin/com/fizzpod/gradle/plugins/layout/LayoutPathSpec.kt
/* (C) 2024 */
/* SPDX-License-Identifier: Apache-2.0 */
package com.fizzpod.gradle.plugins.layout

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.io.TempDir
import java.io.File

class LayoutPathSpec {

    @TempDir
    lateinit var temporaryFolder: File

    @Test
    fun `create path in root folder`() {
        // when:
        val layoutPath = LayoutPath("folder1")

        // then:
        layoutPath.write(temporaryFolder)

        // expect:
        val folder = File(temporaryFolder, "folder1")
        assertTrue(folder.exists(), "Folder should exist")
        assertTrue(folder.isDirectory, "Path should be a directory")
    }

    @Test
    fun `create path with forward slash at start`() {
        // when:
        val layoutPath = LayoutPath("/folder1") // Path for LayoutPath

        // then:
        layoutPath.write(temporaryFolder)

        // expect:
        val folder = File(temporaryFolder, "folder1") // Expected path, / is handled by File constructor
        assertTrue(folder.exists(), "Folder should exist")
        assertTrue(folder.isDirectory, "Path should be a directory")
    }

    @Test
    fun `create path with multiple subfolders`() {
        // when:
        val layoutPath = LayoutPath("/folder1/folder1_1/folder1_1_1") // Path for LayoutPath

        // then:
        layoutPath.write(temporaryFolder)

        // expect:
        val folder = File(temporaryFolder, "folder1/folder1_1/folder1_1_1") // Expected path
        assertTrue(folder.exists(), "Folder structure should exist")
        assertTrue(folder.isDirectory, "Path should be a directory")
    }

    @Test
    fun `create path with file in the way`() {
        // setup:
        val existingFile = File(temporaryFolder, "file1")
        existingFile.writeText("This is a test file.") // Spock's newFile creates a file

        // when:
        // Trying to create a directory where a file with the same name exists
        val layoutPath = LayoutPath("/file1")

        // then:
        // LayoutPath.write calls mkdirs(). If "file1" is a file, mkdirs() on it will fail.
        layoutPath.write(temporaryFolder)

        // expect:
        assertTrue(existingFile.exists(), "Existing file should still exist")
        assertTrue(existingFile.isFile, "Path should remain a file, not converted to directory")
        // Ensure it didn't somehow become a directory
        assertFalse(existingFile.isDirectory, "Path should not be a directory")
    }

    @Test
    fun `create path with file in sub folder position`() {
        // setup:
        val existingFile = File(temporaryFolder, "file1")
        existingFile.writeText("This is an existing file.") // Spock's newFile

        // when:
        // Trying to create a directory structure where a segment of the path ("file1") is already a file.
        val layoutPath = LayoutPath("/file1/folder1_1")

        // then:
        // LayoutPath.write will attempt File(root, "/file1/folder1_1").mkdirs().
        // This should fail because "file1" is a file, so "file1/folder1_1" cannot be created.
        layoutPath.write(temporaryFolder)

        // expect:
        assertTrue(existingFile.exists(), "Existing file 'file1' should still exist")
        assertTrue(existingFile.isFile, "Existing 'file1' should be a file")

        // The Spock test had "file1/folder1", assuming it meant "file1/folder1_1" based on the LayoutPath input
        val newFolder = File(temporaryFolder, "file1/folder1_1")
        assertFalse(newFolder.exists(), "New folder should not be created because a parent path segment is a file")
    }
}
